### DISCUSSION TITLE
`.env` vars suck

### DISCUSSION THREAD

**Discussion by mrnerdhair** [shapeshift/web#1195] [General]
Keeping `.env` vars in sync sucks, and since we don't need them for injecting secrets they're probably not the best setup.

- Especially as we expand our need for configuration doodads by adding new features and integrations syncing multiple build pipelines will become increasingly untenable. For reference, we currently have to worry about at least:
  - Local developer environments syncing `sample.env` with `.env`
  - Cloudflare Pages build settings for app.shapeshift.com
  - Cloudflare Pages build settings for *.web-29e.pages.dev
  - Fleek build settings
  - Maybe still Vercel build settings in some contexts?
  - The build settings on any downstream forks (say, @BitHighlander's KK client)
- It's quite difficult to do automated deployment verification (i.e. to check if the Fleek build is correct and matches the one from GH Actions, so we don't have to trust the security of their build pipeline) unless the env vars a build will be produced with is predictable.
- Using `.env` vars requires a non-trivial amount of support code.
  - `.env` vars are all untyped strings, and have to be coerced into whatever's required by `config.ts`. This works, but it's not exactly a standards-based approach.
  - There's some real contortions in `react-app-rewired.js` to turn them into `env.json` at runtime, and dropping support would eliminate ~100 lines of some of our hardest-to-maintain code.

I suggest we move to a new model based on JSON files; specifically, `env.prod.json` and `env.dev.json`, with contents matching the current set of parameters used by the Cloudflare Pages build. We can sort the keys into canonical form and use a JSON schema can validate the settings at compile-time. If we *do* need to maintain support for pipeline-specific env-var overrides of these default settings packages, that can be done as well... I figure there's probably a package for that.

**Comment by 0xdef1cafe** [comment]
how does this work with fleek and cloudflare?

we want the ability to update environment variables in hosted environments without making application code changes, which this would prevent.

this also goes against best practises of not baking environment variables into builds https://12factor.net/

**Comment by con5cience** [comment]
My 2 cents:
1) Eventually we'll be consolidating on Fleek entirely so the surface for this being a pain in the ass will shrink.
2) We should be setting sane defaults in the code if the environment variables are undefined, otherwise we use the env var values.
3) `dotenv-json` is great for setting env vars from JSON files in NodeJS.

**Reply by 0xdef1cafe** [reply]
> My 2 cents:
> 
> 1. Eventually we'll be consolidating on Fleek entirely so the surface for this being a pain in the ass will shrink.
> 2. We should be setting sane defaults in the code if the environment variables are undefined, otherwise we use the env var values.
> 3. `dotenv-json` is great for setting env vars from JSON files in NodeJS.

ty ser

1. fleek is nice for updating env vars
2. & 3 - we set sane defaults, and have centralized config validation implemented here (similar to dotenv-json) https://github.com/shapeshift/web/blob/develop/src/config.ts#L10
